File: packages/arcis-python/arcis/middleware/cors.py
"""
Arcis Middleware - Safe CORS

Secure CORS handler with safe defaults. Unlike permissive CORS libraries:
- No wildcard `*` when credentials are enabled
- `null` origin is always blocked
- `Vary: Origin` is always set for proper caching
- You must explicitly configure allowed origins
"""

import re
from typing import Callable, List, Optional, Pattern, Sequence, Union

DEFAULT_METHODS = ["GET", "HEAD", "PUT", "PATCH", "POST", "DELETE"]
DEFAULT_HEADERS = ["Content-Type", "Authorization"]
DEFAULT_MAX_AGE = 600


def _is_origin_allowed(
    request_origin: str,
    allowed: Union[str, List[str], "Pattern[str]", Callable[[str], bool], bool],
) -> bool:
    """Check if an origin is allowed by the configured policy."""
    # 'null' origin is always blocked — sent by sandboxed iframes, data: URIs
    if request_origin == "null":
        return False

    if allowed is True:
        return True

    if isinstance(allowed, str):
        return request_origin == allowed

    if isinstance(allowed, list):
        return request_origin in allowed

    if isinstance(allowed, re.Pattern):
        return bool(allowed.match(request_origin))

    if callable(allowed):
        return allowed(request_origin)

    return False


class SafeCors:
    """
    Safe CORS handler.

    Args:
        origin: Allowed origins — string, list, regex, callable, or True (reflect, dev only)
        methods: Allowed HTTP methods
        allowed_headers: Allowed request headers
        exposed_headers: Headers exposed to the browser
        credentials: Allow credentials (cookies, auth headers)
        max_age: Preflight cache duration in seconds

    Example:
        cors = SafeCors(origin="https://myapp.com", credentials=True)
    """

    def __init__(
        self,
        origin: Union[str, List[str], "Pattern[str]", Callable[[str], bool], bool],
        methods: Optional[Sequence[str]] = None,
        allowed_headers: Optional[Sequence[str]] = None,
        exposed_headers: Optional[Sequence[str]] = None,
        credentials: bool = False,
        max_age: int = DEFAULT_MAX_AGE,
    ):
        self.origin = origin
        self.methods = list(methods) if methods else DEFAULT_METHODS
        self.allowed_headers = list(allowed_headers) if allowed_headers else DEFAULT_HEADERS
        self.exposed_headers = list(exposed_headers) if exposed_headers else []
        self.credentials = credentials
        self.max_age = max_age

    def get_headers(self, request_origin: Optional[str], method: str = "GET") -> dict:
        """
        Get CORS response headers for a request.

        Args:
            request_origin: The Origin header from the request (None if absent)
            method: The HTTP method of the request

        Returns:
            Dict of CORS headers to add to the response
        """
        headers: dict = {"Vary": "Origin"}

        # No origin = same-origin request, skip CORS headers
        if not request_origin:
            return headers

        if not _is_origin_allowed(request_origin, self.origin):
            return headers

        headers["Access-Control-Allow-Origin"] = request_origin

        if self.credentials:
            headers["Access-Control-Allow-Credentials"] = "true"

        if self.exposed_headers:
            headers["Access-Control-Expose-Headers"] = ", ".join(self.exposed_headers)

        # Preflight headers
        if method == "OPTIONS":
            headers["Access-Control-Allow-Methods"] = ", ".join(self.methods)
            headers["Access-Control-Allow-Headers"] = ", ".join(self.allowed_headers)
            headers["Access-Control-Max-Age"] = str(self.max_age)

        return headers

    def flask_handler(self, response):
        """
        Flask after_request handler.

        Example:
            cors = SafeCors(origin="https://myapp.com")

            @app.after_request
            def add_cors(response):
                return cors.flask_handler(response)
        """
        from flask import request

        origin = request.headers.get("Origin")
        headers = self.get_headers(origin, request.method)
        for key, value in headers.items():
            response.headers[key] = value

        if request.method == "OPTIONS" and "Access-Control-Allow-Origin" in headers:
            response.status_code = 204

        return response


def create_cors(
    origin: Union[str, List[str], "Pattern[str]", Callable[[str], bool], bool],
    methods: Optional[Sequence[str]] = None,
    allowed_headers: Optional[Sequence[str]] = None,
    exposed_headers: Optional[Sequence[str]] = None,
    credentials: bool = False,
    max_age: int = DEFAULT_MAX_AGE,
) -> SafeCors:
    """
    Create a safe CORS handler.

    Example:
        cors = create_cors(origin=["https://myapp.com", "https://admin.myapp.com"])
    """
    return SafeCors(
        origin=origin,
        methods=methods,
        allowed_headers=allowed_headers,
        exposed_headers=exposed_headers,
        credentials=credentials,
        max_age=max_age,
    )

File: packages/arcis-node/src/middleware/cors.ts
/**
 * @module @arcis/node/middleware/cors
 * Safe CORS middleware with secure defaults
 */

import type { Request, Response, NextFunction, RequestHandler } from 'express';

/** CORS configuration options */
export interface CorsOptions {
  /**
   * Allowed origins. Can be:
   * - A string: exact match (e.g., 'https://example.com')
   * - An array: whitelist of allowed origins
   * - A RegExp: pattern match (use with care)
   * - A function: custom validation `(origin) => boolean`
   * - `true`: reflect the request origin (DANGEROUS — only for dev)
   *
   * Default: none (no origin allowed). You must explicitly set this.
   */
  origin: string | string[] | RegExp | ((origin: string) => boolean) | true;

  /** Allowed HTTP methods. Default: ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE'] */
  methods?: string[];

  /** Allowed headers. Default: ['Content-Type', 'Authorization'] */
  allowedHeaders?: string[];

  /** Headers exposed to the browser. Default: [] */
  exposedHeaders?: string[];

  /** Allow credentials (cookies, authorization headers). Default: false */
  credentials?: boolean;

  /** Preflight cache duration in seconds. Default: 600 (10 minutes) */
  maxAge?: number;

  /** Respond to preflight with 204 (no content). Default: true */
  preflightContinue?: boolean;
}

const DEFAULT_METHODS = ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE'];
const DEFAULT_HEADERS = ['Content-Type', 'Authorization'];
const DEFAULT_MAX_AGE = 600;

/**
 * Check if an origin is allowed by the configured policy.
 */
function isOriginAllowed(
  requestOrigin: string,
  allowed: CorsOptions['origin']
): boolean {
  // 'null' origin is always blocked — sent by sandboxed iframes, data: URIs, etc.
  if (requestOrigin === 'null') return false;

  if (allowed === true) return true;

  if (typeof allowed === 'string') {
    return requestOrigin === allowed;
  }

  if (Array.isArray(allowed)) {
    return allowed.includes(requestOrigin);
  }

  if (allowed instanceof RegExp) {
    return allowed.test(requestOrigin);
  }

  if (typeof allowed === 'function') {
    return allowed(requestOrigin);
  }

  return false;
}

/**
 * Create safe CORS middleware.
 *
 * Unlike permissive CORS libraries, this enforces secure defaults:
 * - No wildcard `*` when credentials are enabled
 * - `null` origin is always blocked
 * - `Vary: Origin` is always set for proper caching
 * - You must explicitly configure allowed origins
 *
 * @param options - CORS configuration
 * @returns Express middleware
 *
 * @example
 * // Allow a single origin
 * app.use(safeCors({ origin: 'https://myapp.com' }));
 *
 * @example
 * // Allow multiple origins with credentials
 * app.use(safeCors({
 *   origin: ['https://myapp.com', 'https://admin.myapp.com'],
 *   credentials: true,
 * }));
 *
 * @example
 * // Development: allow all (NOT for production)
 * app.use(safeCors({ origin: true }));
 */
export function safeCors(options: CorsOptions): RequestHandler {
  const {
    origin,
    methods = DEFAULT_METHODS,
    allowedHeaders = DEFAULT_HEADERS,
    exposedHeaders = [],
    credentials = false,
    maxAge = DEFAULT_MAX_AGE,
    preflightContinue = true,
  } = options;

  return (req: Request, res: Response, next: NextFunction) => {
    const requestOrigin = req.headers.origin;

    // Always set Vary: Origin for proper caching
    res.setHeader('Vary', 'Origin');

    // No origin header = same-origin request, skip CORS headers
    if (!requestOrigin) {
      return next();
    }

    const allowed = isOriginAllowed(requestOrigin, origin);

    if (!allowed) {
      // Don't set any CORS headers — browser will block the request
      return next();
    }

    // Set Access-Control-Allow-Origin to the specific origin (not *)
    res.setHeader('Access-Control-Allow-Origin', requestOrigin);

    if (credentials) {
      res.setHeader('Access-Control-Allow-Credentials', 'true');
    }

    if (exposedHeaders.length > 0) {
      res.setHeader('Access-Control-Expose-Headers', exposedHeaders.join(', '));
    }

    // Handle preflight requests
    if (req.method === 'OPTIONS') {
      res.setHeader('Access-Control-Allow-Methods', methods.join(', '));
      res.setHeader('Access-Control-Allow-Headers', allowedHeaders.join(', '));
      res.setHeader('Access-Control-Max-Age', String(maxAge));

      if (preflightContinue) {
        res.status(204).end();
        return;
      }
    }

    next();
  };
}

/**
 * Alias for safeCors
 * @see safeCors
 */
export const createCors = safeCors;
